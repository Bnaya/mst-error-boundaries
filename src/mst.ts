import { types, typecheck, IAnyModelType } from "mobx-state-tree";

export function addMSTErrorBoundary<IT extends IAnyModelType>(
  Model: IT,
  /**
   * Supply a valid snapshot to use, instead of the failing one
   * We also give you the validation error of the snapshot, and the bad snapshot.
   */
  getFallbackSnapshot: (
    error: unknown,
    originalButBadSnapshot: unknown
  ) => IT["CreationType"]
) {
  return types.snapshotProcessor(Model, {
    preProcessor(snMaybeBad: IT["CreationType"]): IT["CreationType"] {
      try {
        typecheck(Model, snMaybeBad);
        return snMaybeBad;
      } catch (e) {
        return getFallbackSnapshot(e, snMaybeBad);
      }
    }
  });
}

function makeTuple<TARGS extends (string | number)[]>(...args: TARGS) {
  return args;
}

export const Point = addMSTErrorBoundary(
  types.model("Point", {
    x: types.number,
    y: types.number
  }),
  (_error, _originalBadSnapshot) => {
    //console.warn(_error)
    return {
      x: 0,
      y: 0
    };
  }
);

export const Box = addMSTErrorBoundary(
  types.model("Box", {
    point: Point,
    width: types.number,
    height: types.number,
    origin: types.enumeration(
      "Origin",
      makeTuple(
        "top-left",
        "top-right",
        "center",
        "bottom-left",
        "bottom-right"
      )
    )
  }),
  () => ({
    point: { x: 1, y: 1 } as any,
    width: 100,
    height: 100,
    origin: "center" as const
  })
);

export const Scene = addMSTErrorBoundary(
  types.model("Scene", {
    boxes: types.array(Box)
  }),
  () => ({
    boxes: []
  })
);
