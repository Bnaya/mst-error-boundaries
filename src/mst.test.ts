import { getSnapshot } from "mobx-state-tree";
import { Point, Box, Scene } from "./mst";

describe("mst error boundaries test", () => {
  test("basic point falls back to 0 0 ", () => {
    const point = Point.create({} as any);

    expect(getSnapshot(point)).toMatchInlineSnapshot(`
      Object {
        "x": 0,
        "y": 0,
      }
    `);
  });

  test("basic box falls back to ??", () => {
    const box = Box.create({
      height: "333" as any,
      origin: "bottom-left",
      width: 333,
      point: {
        x: "5",
        y: "5"
      } as any
    });

    expect(getSnapshot(box)).toMatchInlineSnapshot(`
            Object {
              "height": 100,
              "origin": "center",
              "point": Object {
                "x": 1,
                "y": 1,
              },
              "width": 100,
            }
        `);
  });

  test("basic Scene falls back to ??", () => {
    const scene = Scene.create({
      boxes: [
        {
          height: 50,
          origin: "bottom-left",
          width: 45,
          point: {
            x: "5",
            y: "5"
          } as any
        },
        {
          height: 50,
          origin: "bottom-left",
          width: 45,
          point: {
            x: "5",
            y: "5"
          } as any
        }
      ]
    });

    expect(getSnapshot(scene)).toMatchInlineSnapshot(`
                              Object {
                                "boxes": Array [
                                  Object {
                                    "height": 50,
                                    "origin": "bottom-left",
                                    "point": Object {
                                      "x": 0,
                                      "y": 0,
                                    },
                                    "width": 45,
                                  },
                                  Object {
                                    "height": 50,
                                    "origin": "bottom-left",
                                    "point": Object {
                                      "x": 0,
                                      "y": 0,
                                    },
                                    "width": 45,
                                  },
                                ],
                              }
                    `);
  });
});
