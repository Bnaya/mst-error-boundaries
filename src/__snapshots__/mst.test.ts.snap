// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mst error boundaries test Play with simple types 1`] = `
Object {
  "num": 1,
}
`;

exports[`mst error boundaries test Play with simple types 2`] = `
Object {
  "num": 2,
}
`;
